src/io_uring.c

```diff
@@ -0,0 +1,97 @@
+#include "io_uring.h"
+
+#ifdef HAVE_LIBURING
+#include <liburing.h>
+#include <string.h>
+
+#include "serverassert.h"
+#include "zmalloc.h"
+
+#define IO_URING_DEPTH 256 /* io_uring instance queue depth. */
+
+static size_t io_uring_queue_len = 0;
+static io_uring *_io_uring;
+
+void initIOUring(void) {
+    struct io_uring_params params;
+    struct io_uring *ring = zmalloc(sizeof(struct io_uring));
+    memset(&params, 0, sizeof(params));
+    /* On success, io_uring_queue_init_params(3) returns 0 and ring will
+     * point to the shared memory containing the io_uring queues.
+     * On failure -errno is returned. */
+    assert(io_uring_queue_init_params(IO_URING_DEPTH, ring, &params) == 0);
+    _io_uring = ring;
+}
+
+io_uring *getIOUring() {
+    if (_io_uring) {
+        return _io_uring;
+    } else {
+        initIOUring();
+        return _io_uring;
+    }
+}
+
+/* Submit fdatasync request to io_uring. */
+void ioUringPrepFsyncAndSubmit(io_uring *io_uring, int fd) {
+    struct io_uring_sqe *sqe = io_uring_get_sqe(io_uring);
+    io_uring_prep_fsync(sqe, fd, IORING_FSYNC_DATASYNC);
+    io_uring_queue_len++;
+    io_uring_submit(io_uring);
+}
+
+void ioUringWaitFsyncBarrier(io_uring *io_uring) {
+    struct io_uring_cqe *cqe;
+    while (io_uring_queue_len) {
+        if (io_uring_wait_cqe(io_uring, &cqe) == 0) {
+            if (cqe->res < 0) {
+                char errormsg[1024];
+                strerror_r(-cqe->res, errormsg, sizeof(errormsg));
+                fprintf(stderr,
+                        "Can't persist AOF for fsync error when the AOF fsync policy is 'always': %s. Exiting...",
+                        errormsg);
+                exit(1);
+            }
+            io_uring_cqe_seen(io_uring, cqe);
+            io_uring_queue_len--;
+        } else {
+            char errormsg[1024];
+            strerror_r(-cqe->res, errormsg, sizeof(errormsg));
+            fprintf(stderr, "Can't persist AOF for fsync error when the AOF fsync policy is 'always': %s. Exiting...",
+                    errormsg);
+            exit(1);
+        }
+    }
+}
+
+void freeIOUring(void) {
+    if (_io_uring) {
+        io_uring_queue_exit(_io_uring);
+        zfree(_io_uring);
+        _io_uring = NULL;
+    }
+}
+#else
+#ifndef UNUSED
+#define UNUSED(V) ((void)V)
```

Comment:
we should create `util.h` to host all these helper macros/inline functions. we could also move `serverLog` there too.

Reply:
We already have a `util.h`, it is mostly used for string utility which depend on `sds.h`.  io_uring doesn't need sds either. 

I think the code refactor work for the fat `server.h`, maybe we can create an umbrella ticket to track it later if you want?

Reply:
Yeah, we can't fix all the refactoring now. We have to live with some of this mess but we can try to not make it worse than it already is. I like the separate refactoring work for log.h.

For the UNUSED macro, I don't think it's very important. I wouldn't mind if we just cast to void `(void)whatever` without using the UNUSED macro.